### PR TITLE
Synonyms Mngmt API: validation on adding synonyms

### PR DESCRIPTION
**Pull request by mayya-sharipova**
PR #95895 implemented PUT request for adding synonyms, 
but left validation for later. This adds simple validation for synonyms that is not 
attached to specific analyzer: only checking that the format 
of a synonym rule is correct.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java [hunk 1]
```diff
@@ -84,19 +80,91 @@ public String id() {
...
+     * Adopted from org.apache.lucene.analysis.synonym.SolrSynonymParser,
+     * with modifications not to invoke analyze method.
+     *
+     * @return null if the synonym rule is valid, otherwise return the error message.
+     */
+    public String validate() {
+        if (synonyms.length() == 0) {
+            return "[synonyms] field can't be empty";
+        }
+        String[] sides = split(synonyms, "=>");
```
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:111]: @carlosdelest This is the copy from `SolrSynonymParser` that will eventually parse synonyms for specific analyzers. I wanted to keep the logic the same. 
When this  `SolrSynonymParser`  uses regex, we can modify this as well

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml [hunk 2]
```diff
@@ -22,3 +22,48 @@
...
+  - do:
+      catch: /More than one explicit mapping specified in the same synonyms rule/
+      synonyms.put:
+        synonyms_set: test-update-synonyms
+        body:
+          synonyms_set:
+            - synonyms: "bye => => goodbye"
+
+  - do:
+      catch: /None of the supplied synonyms can be empty/
```
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml:48]: Thanks for the feedback, I will modify the error message as suggested

File: server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:111]: If the goal is to use the same code, why not use the `SolrSynonymParser.parse()` method? That way we wouldn't be copying internal code from it, but using its public methods...

[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:111]: `SolrSynonymParser.parse()` method works in the context of a specific provided analyzer (where synonyms rules are valid for the provided analyzer). From it, I extracted only pieces for the general validation

[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:111]: `SolrSynonymParser.parse()` method works in the context of a specific provided analyzer (where synonyms rules are valid for the provided analyzer). From it, I extracted only pieces for the general validation

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml]: Address in bd06ed580c009f6678d6de9edb0bf2872da24bbc

File: server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:111]: Thanks for explaining Mayya. I'd prefer not to copy this code or use a more clear implementation, but I won't oppose to it. We can always revisit this later.

**Review by carlosdelest — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymRule.java:111]: @carlosdelest Thanks, I've added TODO item to revisit this later.

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/bwc

**Comment by mayya-sharipova** [comment]
@elasticsearchmachine run elasticsearch-ci/bwc

**Pull request merged**